src/Microsoft.OData.Core/Uri/NodeToStringBuilder.cs

```diff
@@ -948,5 +730,95 @@ private static bool IsValidSearchWord(string text)
                 return true;
             }
         }
+
+        /// <summary>
+        /// Check whether Navigation source of the FilterClause rangeVariable is different from the Expression rangeVariable.
+        /// </summary>
+        /// <param name="node">Expression node.</param>
+        /// <returns>If Navigation Source are different, returns true. Otherwise false.</returns>
+        private bool IsDifferentSource(QueryNode node)
+        {
+            ResourceRangeVariableReferenceNode rangeVariableReferenceNode = GetResourceRangeVariableReferenceNode(node);
+            if (rangeVariableReferenceNode == null)
+            {
+                return false;
+            }
+            if (rangeVariables.Count == 0)
+            {
+                return false;
+            }
+
+            RangeVariable rangeVariable = rangeVariables.Peek();

```

Comment:
Add a comment that this is the filterClause rangeVariable.
Better rename in to `filterClauseRangeVariable`
`RangeVariable filterClauseRangeVariable= rangeVariables.Peek();`

Reply:
In IsDifferentSource, we should be comparing the NavigationSource of the RangeVariableReferenceNode and the filterClause rangeVariable NavigationSource.
Therefore the rangeVariable should not be rangeVariables.Peek() since there are instances this will not be the filterClause rangeVariable